File: module_5/python/max_rectangle.py
"""
Title: Max rectangle

Problem:
    Given a 2D binary matrix filled with 0's and 1's, find the largest
    rectangle containing only 1's and return its area.

Execution: python max_rectangle.py
"""
import unittest


def histogram_rectangle(heights):
    if not heights:
        return 0

    stack = []
    i = 0
    maxArea = 0

    while i < len(heights):
        if not stack:
            stack.append(i)
        else:
            while stack and heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
                l = stack[-1] if stack else -1
                width = i - l - 1
                maxArea = max(maxArea, width * h)
            stack.append(i)
        i += 1

    while stack:
        h = heights[stack.pop()]
        l = stack[-1] if stack else -1
        width = i - l - 1
        maxArea = max(maxArea, width * h)
    return maxArea


def max_rectangle(matrix):
    """
    :type matrix: List[List[str]]
    :rtype: int
    """

    if not matrix or not matrix[0]:
        return 0

    rows = len(matrix)
    cols = len(matrix[0])
    heights = [0] * cols
    maxArea = 0

    for r in range(rows):
        for c in range(cols):
            if matrix[r][c] == '1':
                heights[c] += 1
            else:
                heights[c] = 0
        area = histogram_rectangle(heights)
        maxArea = max(maxArea, area)
    return maxArea


class TestMaxRectangle(unittest.TestCase):
    """Unit test for max_rectangle."""

    def test_1(self):
        test_1_input = [
              ["1", "0", "1", "0", "0"],
              ["1", "0", "1", "1", "1"],
              ["1", "1", "1", "1", "1"],
              ["1", "0", "0", "1", "0"]
            ]
        self.assertEqual(max_rectangle(test_1_input), 6)


if __name__ == '__main__':
    unittest.main()

File: module_5/java/MaxRectangle.java
/*
 *   Title: Max rectangle
 *
 *   Problem:
 *   Given a 2D binary matrix filled with 0's and 1's, find the largest
 *   rectangle containing only 1's and return its area.
 *
 *   Execution: javac MaxRectangle.java && java MaxRectangle
 */
import java.util.*;


public class MaxRectangle {

    public static int maxRectangle(char[][] matrix) {
        if (matrix.length <= 0) {
            return 0;
        }
        int n = matrix.length;
        int m = matrix[0].length;
        int[][] dp = new int[n][m];
        int maxArea = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
            if (i == 0) {
              dp[i][j] = matrix[i][j] == '1' ? 1 : 0;
            }
            else {
              dp[i][j] = matrix[i][j] == '1' ? (dp[i-1][j] + 1) : 0;
            }
            int min = dp[i][j];

            for (int k = j; k >= 0; k--) {
              if (min == 0) {
                  break;
              }
              if (dp[i][k] < min) {
                  min = dp[i][k];
              }
              maxArea = Math.max(maxArea, min * (j - k + 1));
            }
          }
        }
        return maxArea;
  }

    public static void main(String[] args) {
        char[][] matrix = { {'1', '0', '1', '0', '0'},
                            {'1', '0', '1', '1', '1'},  
                            {'1', '1', '1', '1', '1'},
                            {'1', '0', '0', '1', '0'}};
        assert maxRectangle(matrix) == 6;

        System.out.println("Passed all test cases");
    }
}
